core/parachain/availability/chunks.hpp

```diff
@@ -6,74 +6,75 @@
-    unsigned long out = 0;
-    auto r = ECCR_get_recovery_threshold(validators, &out);
-    if (r.tag != NPRSResult_Tag::NPRS_RESULT_OK) {
-      return ErasureCodingError(r.tag);
+    auto res = ec_cpp::getRecoveryThreshold(10);
+    if (ec_cpp::resultHasError(res)) {
+      return ErasureCodingError(ec_cpp::resultGetError(std::move(res)));
     }
-    return out;
+    return ec_cpp::resultGetValue(std::move(res));
   }
 
   inline outcome::result<std::vector<network::ErasureChunk>> toChunks(
       size_t validators, const runtime::AvailableData &data) {
     OUTCOME_TRY(message, scale::encode(data));
-    DataBlock _message{message.data(), message.size()};
-    ChunksList _list{nullptr, 0};
-    auto _free = gsl::finally([&] {
-      if (_list.data) {
-        ECCR_deallocate_chunk_list(&_list);
-      }
-    });
-    auto r = ECCR_obtain_chunks(validators, &_message, &_list);
-    if (r.tag != NPRSResult_Tag::NPRS_RESULT_OK) {
-      return ErasureCodingError(r.tag);
+
+    auto create_result = ec_cpp::create(validators);
+    if (ec_cpp::resultHasError(create_result)) {
+      return ErasureCodingError(
+          ec_cpp::resultGetError(std::move(create_result)));
+    }
+
+    auto encoder = ec_cpp::resultGetValue(std::move(create_result));
+    auto encode_result =
+        encoder.encode(ec_cpp::Slice<uint8_t>(message.data(), message.size()));
+    if (ec_cpp::resultHasError(encode_result)) {
+      return ErasureCodingError(
+          ec_cpp::resultGetError(std::move(encode_result)));
     }
-    assert(_list.count == validators);
+
+    auto shards = ec_cpp::resultGetValue(std::move(encode_result));
+    assert(shards.size() == validators);
+
     std::vector<network::ErasureChunk> chunks;
     chunks.resize(validators);
     for (size_t i = 0; i < validators; ++i) {
-      auto &_chunk = _list.data[i];
-      assert(_chunk.index == i);
       auto &chunk = chunks[i];
       chunk.index = i;
-      chunk.chunk = common::BufferView(_chunk.data.array, _chunk.data.length);
+      chunk.chunk = std::move(shards[i]);
     }
     return chunks;
   }
 
   inline outcome::result<runtime::AvailableData> fromChunks(
       size_t validators, const std::vector<network::ErasureChunk> &chunks) {
-    std::vector<Chunk> _chunks;
+    auto create_result = ec_cpp::create(validators);
+    if (ec_cpp::resultHasError(create_result)) {
+      return ErasureCodingError(
+          ec_cpp::resultGetError(std::move(create_result)));
+    }
+
+    auto encoder = ec_cpp::resultGetValue(std::move(create_result));
+    std::vector<decltype(encoder)::Shard> _chunks;
     _chunks.resize(chunks.size());
     for (size_t i = 0; i < chunks.size(); ++i) {
-      auto &_chunk = _chunks[i];
       auto &chunk = chunks[i];
-      _chunk.index = chunk.index;
-      _chunk.data = {
-          const_cast<uint8_t *>(chunk.chunk.data()),
-          chunk.chunk.size(),
-      };
+      _chunks[chunk.index] = std::move(chunk.chunk);
```

Comment:
Check `chunk.index < validators` before using in `operator[]` (e.g. `chunks=[{index=100}]`)